### ISSUE TITLE
robot-name: test runner fails on bonus test

### ISSUE THREAD

**Issue by bitfield** [exercism/go#1526]
I see from [this solution](https://exercism.org/mentoring/discussions/0688e21fef5847ad808501586468801a) that the test runner is apparently having trouble running the bonus tests (protected by a build tag):

```
We received the following error when we ran your code:
FAIL	robotname [build failed]
'/usr/local/go/bin/go test --json .' returned exit code 2: exit status 2
```

So the tests show as failed even though the student is not trying to pass the bonus test. We need to figure out what to do here, don't we? Perhaps the simplest thing is just to drop the 'bonus' idea and make it part of the standard tests.

For those not familiar with the exercise, the bonus test checks that the solution does not hang when there are no more possible names to generate, but instead returns an error. I think it's quite reasonable to include this as a non-bonus part of the exercise, but what does everybody else think?

**Comment by louis-cl** [comment]
If keeping the simplicity of not having the bonus is important, the bonus test could be extracted as its own new problem.
So the easy version  just focuses on teaching random generation.

Then a harder version like "robot name reloaded" can have the "bonus".
Has such pattern (exercise building on each other) been used before / is desired ?

**Comment by bitfield** [comment]
One problem with the simple version (that is, the one that doesn't detect running out of names) is that the benchmark hangs, and students usually don't understand why this is. (It's because the benchmarker is quite liable to run the function more than 676,000 times, thus exhausting the robot namespace.)

A common way to solve this is by generating all possible names _in advance_ and permuting them, which is desperately slow and wasteful, and complicated.

On the other hand, if you need to pass the bonus test (which is specifically that your function returns an error when there are no more names available), then you can write the simple solution (generate names randomly on request) with no problems. So my vote would be to make the bonus test part of the normal tests, and drop the build tag mechanism (which we don't use for any other exercise, as far as I know).

**Comment by jmrunkle** [comment]
How about we give the student control by putting a constant in `robot_name.go` that defaults to false then the bonus instructions could say set the constant to true to enable the bonus test. Then we put everything in one test file as bitfield suggests and use the constant value in `robot_name.go` to noop the test unless they have enabled it. Allows students to opt into the extra testing without forcing everyone to do the bonus test unless they want to. Seems like a balance of alternative viewpoints and fixes the main problem bitfield identified.

Also, this exercise in general seems to be pretty weak on [instructions](https://github.com/exercism/go/blob/main/exercises/practice/robot-name/.docs/instructions.md) given that we give them [a completely empty stub file](https://github.com/exercism/go/blob/main/exercises/practice/robot-name/robot_name.go) and the tests are not trivially instructive about what the boolean inputs mean (eg. [this test](https://github.com/exercism/go/blob/main/exercises/practice/robot-name/robot_name_test.go#L11)).

**Comment by bitfield** [comment]
Just to be clear, I'm suggesting that the 'doesn't hang forever when out of names' test is a really good way of steering students towards a better solution without mentor input (or perhaps more accurately, steering them away from a variety of bad solutions). This seems more in the spirit of Exercism V3, doesn't it?

**Comment by jmrunkle** [comment]
> Just to be clear, I'm suggesting that the 'doesn't hang forever when out of names' test is a really good way of steering students towards a better solution without mentor input (or perhaps more accurately, steering them away from a variety of bad solutions). This seems more in the spirit of Exercism V3, doesn't it?

I like that. I am sufficiently sold on this idea. So current proposal here:

1. Make the bonus test part of the standard test suite.
2. Update the instructions to be more explicit about this requirement.
3. Remove the Benchmark test since it is pretty redundant once you add the bonus test (and students using the UI will not be able to see the benchmark anyways).

**Comment by jmrunkle** [comment]
@ekingery - do you agree with that plan?

NOTE: we already skip benchmark tests in the test runner (somewhat manually).

**Comment by ekingery** [comment]
> @ekingery - do you agree with that plan?

Yes, I agree. Thanks for checking. The build tags should definitely be a separate concept (if that), and not included as an optional sidebar. I'm pretty sure that is not a pattern we want to encourage.

Overall, it's a tough call because the bonus makes it a much more involved problem. But that's probably fine as it's just a practice exercise and doesn't have stated concepts or goals, etc. I'm guessing the lack of clear instructions and context around why the unique aspect is important or interesting is causing additional confusion / poor solutions.
